src/adaptor/ini.rs

```diff
@@ -0,0 +1,502 @@
+
+use nom::{IResult, space, alphanumeric, multispace};
+
+use super::{Adaptor, Config, AdaptorError};
+
+use std::collections::{HashMap, hash_map};
+use std::fmt::Debug;
+use std::io::{Read, Write};
+use std::iter::FromIterator;
+use std::str;
+
+/// The adaptor struct for INI files
+/// Later, this might contain parameters for the myriad INI quirks
+pub struct IniAdaptor {
+    
+}
+
+impl IniAdaptor {
+    /// Constructs a new `IniAdaptor`
+    pub fn new() -> IniAdaptor {
+        IniAdaptor { }
+    }
+}
+
+impl<'a> Adaptor<'a> for IniAdaptor {
+    /// Deserialize the INI data into the `Config` AST
+    fn deserialize<R>(&self, mut reader: R) -> Result<Config, AdaptorError> 
+        where R: Read 
+    {
+        let mut buffer = Vec::new();
+        reader.read_to_end(&mut buffer);
+
+        // parse the basic INI structure
+        let output = match sections(&buffer) {
+            IResult::Done(_, o) => o,
+            _ => return Err("unable to parse INI data"),
+        };
+
+        let mut combined = HashMap::new();
+
+        // Here we convert the INI into our configuration AST,
+        // performing section and key de-duplication as necessary
+        for (name, pairs) in output {
+            // fetch existing entry or create a new one, deduplicating sections
+            let mut entry = combined.entry(name.into()).or_insert_with(|| HashMap::new());
+            // later, we will need schema data in order to encode type information into the AST
+            // for now, just assume everything is a string
+            let converted = pairs.iter().map(|&(key,value)| (key.to_string(), Config::Text(value.to_string())));
+            insert_all(entry, converted);
+        }
+
+        // wrap it all up in an object
+        let objects = combined.into_iter().map(|(key, value)| (key, Config::Object(value))).collect();
+        Ok(Config::Object(objects))
+    }
+
+    /// Serialize the `Config` AST into INI format
+    fn serialize<W>(&self, config: Config, mut writer: W) -> Result<(), AdaptorError> 
+        where W: Write 
+    {
+        let ini_model = try!(convert_model(config));
+
+        for (header, props) in ini_model {
+            writeln!(writer, "[{}]", header);
+            for (key, value) in props {
+                writeln!(writer, "{} = {}", key, value);
+            }
+            writeln!(writer, "");
+        }
+
+        Ok(())
+    }
+}
+
+#[test]
```

Comment:
This might be a noob Rust question, but should we put all the tests at the bottom of the file, or in another file?